transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/persistence/SQLiteEventStore.java

```diff
@@ -0,0 +1,355 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.android.datatransport.runtime.scheduling.persistence;
+
+import android.content.ContentValues;
+import android.content.Context;
+import android.database.Cursor;
+import android.database.sqlite.SQLiteDatabase;
+import android.database.sqlite.SQLiteOpenHelper;
+import android.support.annotation.Nullable;
+import com.google.android.datatransport.Priority;
+import com.google.android.datatransport.runtime.EventInternal;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Set;
+import javax.inject.Inject;
+
+/** {@link EventStore} implementation backed by a SQLite database. */
+public class SQLiteEventStore implements EventStore {
+
+  static final int MAX_RETRIES = 10;
+
+  private static final Priority[] ALL_PRIORITIES = Priority.values();
+
+  private final OpenHelper openHelper;
+  private SQLiteDatabase db;
+
+  @Inject
+  SQLiteEventStore(Context applicationContext) {
+    this.openHelper = new OpenHelper(applicationContext);
+  }
+
+  private SQLiteDatabase getDb() {
+    if (db == null) {
+      db = openHelper.getWritableDatabase();
+    }
+    return db;
+  }
+
+  @Override
+  public <T> T atomically(AtomicFunction<T> function) {
+    return inTransaction(db -> function.execute());
+  }
+
+  @Override
+  public PersistedEvent persist(String backendName, EventInternal event) {
+    ContentValues values = new ContentValues();
+    values.put("backend_id", backendName);
+    values.put("transport_name", event.getTransportName());
+    values.put("priority", event.getPriority().ordinal());
+    values.put("timestamp_ms", event.getEventMillis());
+    values.put("uptime_ms", event.getUptimeMillis());
+    values.put("payload", event.getPayload());
+    values.put("num_attempts", 0);
+
+    long newRowId =
+        inTransaction(
```

Comment:
I thought the idea was that callers would call something like

```
store.atomically({
  store.persist(foo, bar);
})
```

Doesn't that mean you're creating a nested transaction by default in this usage?

If atomically is required, persist and friends should assert that they're called in the block rather than opening a nested transaction.

A type safe way to make this work is to have persist and similar writing methods on a nested type like EventStore.Writer and have atomically pass that in.

Then you get

```
store.atomically(writer -> {
  writer.persist(foo, bar);
})
```

This would also make it easy to have a thread-local writer that would allow concurrent usage.

Reply:
Fair enough.

Reading a bit more about this, it seems like there's [no real penalty to using nested transactions this way](https://stackoverflow.com/questions/26106237/sqlitedatabase-nested-transaction-and-workaround) (and no benefit either) so it's not harmful.